### ISSUE TITLE
Index out of bounds error encountered when parsing large fields

### ISSUE THREAD

**Issue by Volleygram** [JoshClose/CsvHelper#1674]
Version 21.1.0. Index out of bounds error encountered when parsing large fields.

In CsvParser.cs, line 752, a check is performed to allocate more memory for processing larger fields. If the current allocation is insufficient, the buffer size is doubled. 

			if (newLength > processFieldBuffer.Length)
			{
				processFieldBufferSize *= 2;
				ArrayPool<char>.Shared.Return(processFieldBuffer);
				processFieldBuffer = ArrayPool<char>.Shared.Rent(processFieldBufferSize);
			}

The issue is that is that at times, the doubled size can still insufficient and an out of bounds exception is throw on line 782.

In my particular case, the newLength calculated is 5953, so the processedFieldBuffer is expanded 2048 to 4096. However that is still insufficient.

I would expect through a while loop or other mechanism, the memory allocation would continue to grow until it is sufficient to hold the processed body of text.

**Comment by JoshClose** [comment]
Changing to this:

```cs
if (newLength > processFieldBuffer.Length)
{
	while (newLength > processFieldBufferSize)
	{
		processFieldBufferSize *= 2;
	}

	processFieldBuffer = new char[processFieldBufferSize];
}
```

**Comment by JoshClose** [comment]
Fixed in 21.2.1 in NuGet. https://www.nuget.org/packages/CsvHelper/21.2.1
